commit/merkleroot/rmn/client.go

```diff
@@ -19,13 +33,533 @@ type Client interface {
+) ([]*rmnpb.EcdsaSignature, error) {
+	tReportsInitialRequest := time.NewTimer(c.reportsInitialRequestTimerDuration)
+	reportSigs := make([]*rmnpb.ReportSignature, 0)
+	finishedRequests := mapset.NewSet[uint64]()
+	respChan := c.rawRmnClient.Recv()
+
+	for {
+		select {
+		case resp := <-respChan:
+			responseTyp, err := c.parseResponse(&resp, requestIDs, finishedRequests)
+			if err != nil {
+				c.lggr.Errorw("failed to parse RMN response", "err", err)
+				continue
+			}
+
+			reportSig := responseTyp.GetReportSignature()
+			if reportSig == nil {
+				c.lggr.Errorf("RMN node returned an unexpected type of response: %+v", responseTyp.Response)
+			} else {
+				reportSigs = append(reportSigs, reportSig)
+			}
+
+			if len(reportSigs) >= c.minSigners {
+				ecdsaSigs := make([]*rmnpb.EcdsaSignature, 0)
+				for _, rs := range reportSigs {
+					ecdsaSigs = append(ecdsaSigs, rs.Signature)
+				}
+				return ecdsaSigs, nil
+			}
+		case <-tReportsInitialRequest.C:
+			// Send to the rest of the signers
+			for node := range randomIter(c.rmnNodes) {
+				if !node.IsSigner || signersRequested.Contains(node.ID) {
+					continue
+				}
+				req := &rmnpb.Request{
+					RequestId: newRequestID(),
+					Request: &rmnpb.Request_ReportSignatureRequest{
+						ReportSignatureRequest: reportSigReq,
+					},
+				}
+				if err := c.marshalAndSend(req, node.ID); err != nil {
+					c.lggr.Errorw("failed to send report signature request", "node_id", node.ID, "err", err)
+					continue
+				}
+				requestIDs.Add(req.RequestId)
+				signersRequested.Add(node.ID)
+			}
+		case <-ctx.Done():
+			return nil, ErrTimeout
+		}
+	}
+}
+
+func (c *PBClient) ensureEnoughSignedObservations(rmnSignedObservations []rmnSignedObservationWithMeta) error {
+	counts := make(map[uint64]int)
+	for _, so := range rmnSignedObservations {
+		for _, lu := range so.SignedObservation.Observation.FixedDestLaneUpdates {
+			counts[lu.LaneSource.SourceChainSelector]++
+		}
+	}
+
+	for chain, count := range counts {
+		if count < c.minObservers {
+			return fmt.Errorf("not enough observations for chain=%d count=%d minObservers=%d",
+				chain, count, c.minObservers)
+		}
+	}
+
+	return nil
+}
+
+type RawRmnClient interface {
+	Send(rmnNodeID uint32, request []byte) error
+	// Recv returns a channel which can be used to listen on for
+	// responses by all RMN nodes. This is expected to be monitored
+	// by the plugin in order to get RMN responses.
+	Recv() <-chan RawRmnResponse
+}
+
+type RawRmnResponse struct {
+	RMNNodeID uint32
+	Body      []byte // pb
+}
+
+type lurWithMeta struct {
+	lur      *rmnpb.FixedDestLaneUpdateRequest
+	rmnNodes mapset.Set[uint32]
+}
+
+type rmnSignedObservationWithMeta struct {
+	SignedObservation *rmnpb.SignedObservation
+	RMNNodeID         uint32
+}
+
+func randomIter[T any](s []T) chan T {
+	ch := make(chan T)
+	go func() {
+		defer close(ch)
+		perm := rand.Perm(len(s))
+		for _, i := range perm {
+			ch <- s[i]
+		}
+	}()
+	return ch
+}
```

Comment:
This function seems overly complex. Doesn't it just shuffle an array? Why create channels and goroutines? Why not just use `rand.Perm` directly in the loop and access that particular index in the slice?

```golang
for _, idx := range rand.Perm(len(s)) {
  // check if rmnNode[idx] has already been used, if so, skip
  // else, use
}
```